### PR TITLE
quiltctl: Explictly request PTY allocation in `exec`

### PR DESCRIPTION
**Pull request by kklin**
Before, the `exec` subcommand would failed when run in contexts that
couldn't allocate a pseudo-terminal:
```
# This failed unless we enable psudo-terminal allocation with `-t`.
% ssh localhost quilt exec 1 cat /etc/hosts
```

### CODE REVIEW

**Comment by tejasmanohar** [comment]
or, should we just add a `-t` flag for PTY like docker exec does?

**Comment by kklin** [comment]
@tejasmanohar I think setting it automatically is easier for users.. although I could be missing a reason why Docker doesn't just do it automatically too.

**Comment by tejasmanohar** [comment]
@kklin I think it's just a precaution. Commands you run via `<docker/quilt> exec` may act differently when run in a pseudo-terminal or not (e.g. different output format).

**Comment by kklin** [comment]
@tejasmanohar Good point, I'll make us use a flag then.

**Comment by kklin** [comment]
@tejasmanohar ready for review

File: Makefile [hunk 1]
```diff
@@ -32,7 +32,7 @@ COV_SKIP= /minion/pb /minion/pprofile /api/pb /constants /scripts /quilt-tester
 		  /quilt-tester/tests/100-basic /quilt-tester/tests/100-basic/check_docker.go \
 		  /quilt-tester/tests/100-basic/check_logs.go \
 		  /quilt-tester/tests/spark /quilt-tester/tests/spark/check_spark.go \
-		  /quiltctl/testutils /cluster/provider/mocks
+		  /quiltctl/testutils /cluster/provider/mocks /api/client/mock
```
Comment (tejasmanohar) [inline Makefile]: may as well name the folder "mocks" everywhere to be consistent

File: api/client/getter/getter.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+// New returns an implementation of the Getter interface.
+func New() client.Getter {
+	return clientGetterImpl{addrClientGetterImpl{}}
+}
+
+// We create a separate interface for getting a client given an address so that
+// ContainerClient and LeaderClient can be unit tested.
+type addrClientGetter interface {
```
Comment (tejasmanohar) [inline api/client/getter/getter.go]: not super happy about the concept of a client getter interface but don't have a better drop-in idea rn

**Comment by tejasmanohar** [comment]
(merge conflicts btw)

**Comment by kklin** [comment]
Ready for @ejj review.

**Comment by tejasmanohar** [comment]
Maybe rename PR title

**Comment by tejasmanohar** [comment]
oh, I meant the commit message & PR should probably mention the `-t` flag

**Comment by kklin** [comment]
@tejasmanohar Oh, that makes a lot more sense :P

File: api/client/getter/getter.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	client.Client, error) {
+
+	machines, err := localClient.QueryMachines()
+	if err != nil {
+		return nil, fmt.Errorf("unable to query machines: %s", err.Error())
+	}
+
+	// Try to figure out the lead minion's IP by asking each of the machines
+	// tracked by the local daemon.
+	for _, m := range machines {
```
Comment (ejj) [inline api/client/getter/getter.go]: Could we just check the master machines?

File: api/client/getter/getter.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		log.WithError(err).Debug("Unable to get leader IP.")
+	}
+
+	return nil, errors.New("no leader found")
+}
+
+func (getter clientGetterImpl) ContainerClient(localClient client.Client,
+	stitchID int) (client.Client, error) {
+
+	leaderClient, err := getter.LeaderClient(localClient)
```
Comment (ejj) [inline api/client/getter/getter.go]: Btw, I think we should be able to ask any master where the containers are scheduled not just the leader.  Don't know if that actually works, but it might make sense to implement it, should be trivial.

File: api/util/util.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package util
```
Comment (ejj) [inline api/util/util.go]: Meh, I don't much l like util packages, could this go somewhere else?

That said, I don't feel that strongly about it, if you think this is the best place for it, then fine.

File: quiltctl/command/exec.go [hunk 6]
```diff
@@ -73,6 +78,11 @@ func (eCmd *Exec) Parse(args []string) error {
 
 // Run finds the target continer, and executes the given command in it.
 func (eCmd *Exec) Run() int {
+	if eCmd.allocatePTY && !isTerminal() {
+		log.Error("Cannot allocate pseudo-terminal without a terminal.")
```
Comment (ejj) [inline quiltctl/command/exec.go]: No periods in log messages

File: quiltctl/command/exec.go [hunk 7]
```diff
@@ -100,16 +110,27 @@ func (eCmd *Exec) Run() int {
...
+		if err = eCmd.SSHClient.RequestPTY(); err != nil {
+			log.WithError(err).Info("Error requesting pseudo-terminal")
+			return 1
+		}
 	}
 
-	command := fmt.Sprintf("docker exec -it %s %s", container.DockerID, eCmd.command)
+	var flags string
+	if eCmd.allocatePTY {
+		flags = "-it"
```
Comment (ejj) [inline quiltctl/command/exec.go]: Seems a bit suspicious to me that docker treats `i` and `t` separately, but we're combining them is this the right long term strategy, or a temporary thing?

File: api/client/getter/getter.go (hunk 3 above)
Comment (kklin) [inline api/client/getter/getter.go]: Just to confirm, just check whether the master machine is the leader?

File: api/util/util.go (hunk 5 above)
Comment (kklin) [inline api/util/util.go]: It's used both by `api/client/getter` and `quiltctl/command`. Not sure where else to put it, not against moving it though.

File: api/client/getter/getter.go (hunk 3 above)
Comment (ejj) [inline api/client/getter/getter.go]: Actually I don't think my comment makes sense, nevermind.

File: api/util/util.go (hunk 5 above)
Comment (ejj) [inline api/util/util.go]: fine where it is

File: api/client/getter/getter.go (hunk 4 above)
Comment (kklin) [inline api/client/getter/getter.go]: @ejj would that be cleaner if it was in a separate commit? This commit is just moving the old implementation into an interface.

Comment (kklin) [inline api/client/getter/getter.go]: Hmm now that I look into it more, we only use the `LeaderClient` to query the cluster's container information. So I think this is better done in a separate PR where we get rid of `LeaderClient` altogether. It'd simplify the logic quite a bit. I don't really want to do it in this series though as this is blocking some other patches, and the changes would be done in a separate commit anyways. What do you think?
